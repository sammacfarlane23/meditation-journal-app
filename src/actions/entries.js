import database from '../firebase/firebase';

export const createEntry = (entry) => ({
  type: 'CREATE_ENTRY',
  entry,
});

export const startCreateEntry = (entryData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const { entryText = '', createdAt = 0, title = '' } = entryData;
    const entry = {
      entryText,
      createdAt,
      title,
    };

    return database
      .ref(`users/${uid}/entries`)
      .push(entry)
      .then((ref) => {
        dispatch(
          createEntry({
            id: ref.key,
            ...entry,
          })
        );
      });
  };
};

export const removeEntry = ({ id } = {}) => ({
  type: 'REMOVE_ENTRY',
  id,
});

export const startRemoveEntry = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database
      .ref(`users/${uid}/entries/${id}`)
      .remove()
      .then(() => {
        dispatch(removeEntry({ id }));
      });
  };
};

export const editEntry = (id, updates) => ({
  type: 'EDIT_ENTRY',
  id,
  updates,
});

export const startEditEntry = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database
      .ref(`users/${uid}/entries/${id}`)
      .update(updates)
      .then(() => {
        dispatch(editEntry(id, updates));
      });
  };
};

export const setEntries = (entries) => {
  return {
    type: 'SET_ENTRIES',
    entries,
  };
};

export const startSetEntries = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database
      .ref(`users/${uid}/entries`)
      .once('value')
      .then((snapshot) => {
        const entries = [];

        snapshot.forEach((childSnapshot) => {
          entries.push({
            id: childSnapshot.key,
            ...childSnapshot.val(),
          });
        });
        dispatch(setEntries(entries));
      });
  };
};

export const selectEntryForEdit = (entry) => {
  return {
    type: 'SELECT_ENTRY_FOR_EDIT',
    entry,
  };
};
