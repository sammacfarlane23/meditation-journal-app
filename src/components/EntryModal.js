import React from 'react';
import ReactModal from 'react-modal';
import { connect } from 'react-redux';
import EntryForm from './EntryForm';

export const EntryModal = (props) => (
  <ReactModal
    className='modal'
    isOpen={props.showModal}
    onRequestClose={props.closeModal}
    contentLabel='Entry View'
    closeTimeoutMS={300}
  >
    <EntryForm
      closeModal={props.closeModal}
      entry={props.selectedEntry}
      onSubmit={props.onSubmit}
      canEdit={props.canEdit}
      onRemove={props.onRemove}
    />
  </ReactModal>
);

const mapStateToProps = (state, props) => {
  return {
    entry: state.entries.find((entry) => entry.id === props.id),
  };
};

export default connect(mapStateToProps)(EntryModal);
