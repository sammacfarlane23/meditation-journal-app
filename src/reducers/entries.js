const entriesReducerDefaultState = [];

export default (state = entriesReducerDefaultState, action) => {
  switch (action.type) {
    case 'CREATE_ENTRY':
      return [...state, action.entry];
    case 'REMOVE_ENTRY':
      return state.filter(({ id }) => action.id !== id);
    case 'EDIT_ENTRY':
      return state.map((entry) => {
        if (entry.id === action.id) {
          return {
            ...entry,
            ...action.updates,
          };
        } else {
          return entry;
        }
      });
    case 'SET_ENTRIES':
      return action.entries;
    case 'SELECT_ENTRY_FOR_EDIT':
      return action.entry;
    default:
      return state;
  }
};
