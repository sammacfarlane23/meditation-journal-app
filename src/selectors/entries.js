export default (entries, { text }) => {
  return entries
    .filter((entry) => {
      const titleMatch = entry.title.toLowerCase().includes(text.toLowerCase());

      const textMatch = entry.entryText
        .toLowerCase()
        .includes(text.toLowerCase());

      return textMatch || titleMatch;
    })
    .sort((a, b) => {
      return a.createdAt < b.createdAt ? 1 : -1;
    });
};
