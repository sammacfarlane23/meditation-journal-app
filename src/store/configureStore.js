import { createStore, combineReducers, applyMiddleware } from "redux";
import thunk from "redux-thunk";
import authReducer from "../reducers/auth";
import entriesReducer from "../reducers/entries";
import filtersReducer from "../reducers/filters";

const configuredStore = () => {
  const store = createStore(
    combineReducers({
      entries: entriesReducer,
      auth: authReducer,
      filters: filtersReducer,
    }),
    applyMiddleware(thunk)
  );

  return store;
};

export default configuredStore;
