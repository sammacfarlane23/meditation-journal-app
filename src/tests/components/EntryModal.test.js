import { mount, render } from '@testing-library/react';
import React from 'react';
import EntryModal from '../../components/EntryModal';
import entries from '../fixtures/entries';
import { mountWithReduxState } from '../store/mockStore';

test('should render EntryModal correctly', () => {
  const component = render(<EntryModal />);
  expect(component).toMatchSnapshot();
});

test('should render EntryForm with entry data passed down as prop', () => {
  const component = mountWithReduxState(
    <EntryModal selectedEntry={{ ...entries[0] }} />
  );
  expect(component).toMatchSnapshot();
});
