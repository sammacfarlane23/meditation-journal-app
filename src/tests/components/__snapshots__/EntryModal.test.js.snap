// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EntryForm with entry data passed down as prop 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Component
    selectedEntry={
      Object {
        "createdAt": 0,
        "entryText": "This is an entry",
        "id": "1",
        "title": "",
      }
    }
  >
    <Modal
      ariaHideApp={false}
      bodyOpenClassName="ReactModal__Body--open"
      className="modal"
      closeTimeoutMS={300}
      contentLabel="Entry View"
      isOpen={false}
      onRequestClose={undefined}
      parentSelector={[Function]}
      portalClassName="ReactModalPortal"
      role="dialog"
      shouldCloseOnEsc={true}
      shouldCloseOnOverlayClick={true}
      shouldFocusAfterRender={true}
      shouldReturnFocusAfterClose={true}
    >
      <Portal
        containerInfo={
          <div
            class="ReactModalPortal"
          />
        }
      >
        <ModalPortal
          ariaHideApp={false}
          bodyOpenClassName="ReactModal__Body--open"
          className="modal"
          closeTimeoutMS={300}
          contentLabel="Entry View"
          defaultStyles={
            Object {
              "content": Object {
                "WebkitOverflowScrolling": "touch",
                "background": "#fff",
                "border": "1px solid #ccc",
                "borderRadius": "4px",
                "bottom": "40px",
                "left": "40px",
                "outline": "none",
                "overflow": "auto",
                "padding": "20px",
                "position": "absolute",
                "right": "40px",
                "top": "40px",
              },
              "overlay": Object {
                "backgroundColor": "rgba(255, 255, 255, 0.75)",
                "bottom": 0,
                "left": 0,
                "position": "fixed",
                "right": 0,
                "top": 0,
              },
            }
          }
          isOpen={false}
          onRequestClose={undefined}
          parentSelector={[Function]}
          portalClassName="ReactModalPortal"
          role="dialog"
          shouldCloseOnEsc={true}
          shouldCloseOnOverlayClick={true}
          shouldFocusAfterRender={true}
          shouldReturnFocusAfterClose={true}
          style={
            Object {
              "content": Object {},
              "overlay": Object {},
            }
          }
        />
      </Portal>
    </Modal>
  </Component>
</Provider>
`;

exports[`should render EntryModal correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div />
    <div
      class="ReactModalPortal"
    />
  </body>,
  "container": <div />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
